ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -0,0 +1,267 @@
+  , connectionNumber = connectionNumber state + 1
+  }
+  where
+  !numConn = NumberedConnection
+    { number     = connectionNumber state
+    , connection = conn
+    }
+
+removeConnection
+  :: ( Ord identifier )
+  => identifier
+  -> State identifier handle
+  -> State identifier handle
+removeConnection identifier state =
+  let alteration mentry = case mentry of
+        -- removeConnection is called by the thread that was spawned by the
+        -- handler before insertConnection happened for this identifier. If
+        -- there's nothing here then there's a bug in this module.
+        Nothing -> error "bug: entry vanished"
+        Just _  -> Nothing
+  in  state { connectionMap = Map.alter alteration identifier (connectionMap state) }
+
+-- | Generic concurrent connection manager in which at most one connection for
+-- a given identifier is allowed. New connections will re-use existing
+-- connections, and existing connections will be rejected if there is already
+-- one here.
+--
+-- For example, we could choose
+--   resource := Socket
+--   connectionId := exists sockType . (SockAddr sockType, SockAddr sockType)
+-- and this would guarantee at most one connection per socket address pair,
+-- when the sockType is IPv* and it's a TCP socket.
+-- However, for Unix domain sockets, this is no good, because unnamed sockets
+-- all have (==) SockAddr values. A richer connection identifier type would be
+-- needed, and any servers or clients calling into this concurrent connection
+-- manager with Unix domain sockets would be required to come up with unique
+-- identifiers for unnamed ones (probably by taking the file descriptor number).
+concurrent
+  :: forall connectionId resource request reject handle t .
+     ( Ord connectionId )
+  => (forall provenance .
+         Initiated provenance
+      -> connectionId
+      -> resource
+      -> request provenance
+      -> IO (Decision provenance reject handle)
+     )
+  -- ^ A callback to run for each connection. The `handle` gives an interface
+  -- to that connection, allowing for inspection and control etc. of whatever
+  -- it's doing.
+  -- If this throws an exception, it is re-thrown, so that calls to
+  -- `include` will get them.
+  -- Non-exceptional domain-specific reasons to reject a connection are given
+  -- in the rejection variant.
+  -> (Connections connectionId resource request (Reject reject) (Accept handle) IO -> IO t)
+  -> IO t
+concurrent withResource k = do
+  stateVar :: MVar (State connectionId handle) <- newMVar $ State
+    { connectionMap    = Map.empty
+    , connectionNumber = 0
+    }
+  let connections = Connections { include = includeOne stateVar }
+  k connections `finally` killThreads stateVar
+
+  where
+
+  -- Exception handling in here should ensure that if the handler is succesfully
+  -- created, then it ends up in the shared state.
+  --
+  -- FIXME this one definition is too complex. Try to factor it into simpler
+  -- pieces.
+  includeOne
```

Comment:
It's de-facto a smart constructor for `Connections`, maybe refactor it into
```
connections :: MVar (State connectionId handle) -> Connections id resource (Reject reject) (Accept handle) IO
```
 
Another question: do we need `handle` and `resource`, or can we simplify it by just using one (e.g. `handle ~ resource`).

Reply:
It's supposed to be a handle to that resource. In general it's not a good idea to have them be the same. The handle should be a more restricted, higher-level interface to  the resource. If we just gave back the resource then, for example, if it were a socket or file, the caller could close it, but that shouldn't be allowed.

Reply:
I see, this makes sense; what kind of operations on the resource a handle should expose? Probably just getting it's identifier or the identifier of remote peer aka `getSocketName`, `getPeerName`.

Reply:
In a patch that I'm going to push hopefully later today, I have set the handle to be

```Haskell
data ConnectionHandle m = ConnectionHandle
  { status :: STM m ConnectionResult }

data ConnectionResult = Running | Finished (Maybe SomeException)

waitForConnection :: MonadSTM m => ConnectionHandle m -> m (Maybe SomeException)
```

I don't plan on exposing `getSocketName`, there's no good reason why a user should need to know it's a socket (we'll see).

Eventually we'll need to extend it to support the required P2P `PeerActions` record. Things like "promote this connection from warm to hot".